### PR TITLE
[SPARK-17982][SQL] SQLBuilder should wrap the generated SQL with parenthesis for LIMIT

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Currently, `SQLBuilder` handles `LIMIT` by always adding `LIMIT` at the end of the generated subSQL. It makes `RuntimeException`s like the following. This PR adds a parenthesis always except `SubqueryAlias` is used together with `LIMIT`.

**Before**

``` scala
scala> sql("CREATE TABLE tbl(id INT)")
scala> sql("CREATE VIEW v1(id2) AS SELECT id FROM tbl LIMIT 2")
java.lang.RuntimeException: Failed to analyze the canonicalized SQL: ...
```

**After**

``` scala
scala> sql("CREATE TABLE tbl(id INT)")
scala> sql("CREATE VIEW v1(id2) AS SELECT id FROM tbl LIMIT 2")
scala> sql("SELECT id2 FROM v1")
res4: org.apache.spark.sql.DataFrame = [id2: int]
```

**Fixed cases in this PR**

The following two cases are the detail query plans having problematic SQL generations.

1. `SELECT * FROM (SELECT id FROM tbl LIMIT 2)`

    Please note that **FROM SELECT** part of the generated SQL in the below. When we don't use '()' for limit, this fails.

```scala
# Original logical plan:
Project [id#1]
+- GlobalLimit 2
   +- LocalLimit 2
      +- Project [id#1]
         +- MetastoreRelation default, tbl

# Canonicalized logical plan:
Project [gen_attr_0#1 AS id#4]
+- SubqueryAlias tbl
   +- Project [gen_attr_0#1]
      +- GlobalLimit 2
         +- LocalLimit 2
            +- Project [gen_attr_0#1]
               +- SubqueryAlias gen_subquery_0
                  +- Project [id#1 AS gen_attr_0#1]
                     +- SQLTable default, tbl, [id#1]

# Generated SQL:
SELECT `gen_attr_0` AS `id` FROM (SELECT `gen_attr_0` FROM SELECT `gen_attr_0` FROM (SELECT `id` AS `gen_attr_0` FROM `default`.`tbl`) AS gen_subquery_0 LIMIT 2) AS tbl
```

2. `SELECT * FROM (SELECT id FROM tbl TABLESAMPLE (2 ROWS))`

    Please note that **((~~~) AS gen_subquery_0 LIMIT 2)** in the below. When we use '()' for limit on `SubqueryAlias`, this fails.

```scala
# Original logical plan:
Project [id#1]
+- Project [id#1]
   +- GlobalLimit 2
      +- LocalLimit 2
         +- MetastoreRelation default, tbl

# Canonicalized logical plan:
Project [gen_attr_0#1 AS id#4]
+- SubqueryAlias tbl
   +- Project [gen_attr_0#1]
      +- GlobalLimit 2
         +- LocalLimit 2
            +- SubqueryAlias gen_subquery_0
               +- Project [id#1 AS gen_attr_0#1]
                  +- SQLTable default, tbl, [id#1]

# Generated SQL:
SELECT `gen_attr_0` AS `id` FROM (SELECT `gen_attr_0` FROM ((SELECT `id` AS `gen_attr_0` FROM `default`.`tbl`) AS gen_subquery_0 LIMIT 2)) AS tbl
```

## How was this patch tested?

Pass the Jenkins test with a newly added test case.

### CODE REVIEW

**Comment by gatorsmile** [comment]
Wrong JIRA number. : )

**Comment by dongjoon-hyun** [comment]
Ooops. Thank you for fixing me! @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 1]
```diff
@@ -138,9 +138,12 @@ class SQLBuilder private (
     case g: Generate =>
       generateToSQL(g)
 
-    case Limit(limitExpr, child) =>
+    case Limit(limitExpr, child @ SubqueryAlias(_, _, _)) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: shouldn't we create a test in the sql generation test there?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @rxin .
Here, `CREATE VIEW` is not supported by SQLGeneration test suite. 

CREATE VIEW **v1(id2)** AS SELECT id FROM tbl **LIMIT 2**

The case is a view with column names.

**Comment by rxin** [comment]
but you can construct a query with limit to make sure the output has ( ) no?

**Comment by dongjoon-hyun** [comment]
Sorry, but what do you mean by `the output has ( ) no?`?

**Comment by dongjoon-hyun** [comment]
Ah. I see. `( )` means parenthesis, literally. I'll replace `CREATE VIEW` test case with SQL Generation suite. BTW, at the last commit, the PR is changed to add parenthesis if the child of `Limit` is `Project`.

File: sql/hive/src/test/resources/sqlgen/limit.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+-- This file is automatically generated by LogicalPlanToSQLSuite.
+SELECT * FROM (SELECT id FROM tbl LIMIT 2)
+--------------------------------------------------------------------------------
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/resources/sqlgen/limit.sql]: Without this PR, this test case fails because it generate the following query. Note that the third line.
```
SELECT `gen_attr_0` AS `id`
FROM (SELECT `gen_attr_0`
             FROM SELECT `gen_attr_0`
                         FROM (SELECT `id` AS `gen_attr_0`, `name` AS `gen_attr_1`
                                     FROM `default`.`tbl`) AS gen_subquery_0 LIMIT 2) AS tbl
```

**Comment by dongjoon-hyun** [comment]
@rxin . I moved the test case and fixed a bug in `LogicalPlanToSQLSuite`.

**Comment by dongjoon-hyun** [comment]
The current running build has one testcase failure, `when schema inference is turned on, should read partition data`. It seems to be irrelevant.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 3]
```diff
@@ -45,7 +45,7 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
 
   // Used for generating new query answer files by saving
   private val regenerateGoldenFiles: Boolean = System.getenv("SPARK_GENERATE_GOLDEN_FILES") == "1"
-  private val goldenSQLPath = getTestResourcePath("sqlgen")
+  private val goldenSQLPath = "src/test/resources/sqlgen/"
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Hi, @rxin . 
IMO, this bug on `LogicalPlanToSQLSuite` is needed to be fixed soon. If this PR is needed to be reviewed more, should I make another PR containing that only?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 4]
```diff
@@ -138,6 +138,9 @@ class SQLBuilder private (
     case g: Generate =>
       generateToSQL(g)
 
+    case Limit(limitExpr, child @ Project(_, _)) =>
+      s"(${toSQL(child)} LIMIT ${limitExpr.sql})"
+
     case Limit(limitExpr, child) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: is it always safe to just add limit to any plan?

**Comment by dongjoon-hyun** [comment]
@rxin .
I'll add more testcases for this and figure out the correct answer for your question.
> is it always safe to just add limit to any plan?

**Comment by rxin** [comment]
@dongjoon-hyun any update on this?

**Comment by dongjoon-hyun** [comment]
Sorry for delay, @rxin . I'm still working on, and will update in a few days.

**Comment by rxin** [comment]
@dongjoon-hyun we are going to cut the release branch today (or tomorrow) so it would be good to get this in asap.

**Comment by dongjoon-hyun** [comment]
Oh, 2.1-rc1 today? I see.

**Comment by rxin** [comment]
Not rc, but the branch cut.

**Comment by dongjoon-hyun** [comment]
Ah, I see. It's feature freeze. Thank you for informing me.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 5]
```diff
@@ -138,9 +138,12 @@ class SQLBuilder private (
     case g: Generate =>
       generateToSQL(g)
 
-    case Limit(limitExpr, child) =>
+    case Limit(limitExpr, child: SubqueryAlias) =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: `TableSample` like the following makes this plan.
```sql
select * from (select id from tbl tablesample (2 rows))
```

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 6]
```diff
@@ -138,9 +138,12 @@ class SQLBuilder private (
...
       s"${toSQL(child)} LIMIT ${limitExpr.sql}"
 
+    case Limit(limitExpr, child) =>
+      s"(${toSQL(child)} LIMIT ${limitExpr.sql})"
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: For all child logical plan except `SubqueryAlias`, `()` is used.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 7]
```diff
@@ -162,6 +165,8 @@ class SQLBuilder private (
     case p: Except =>
       build("(" + toSQL(p.left), ") EXCEPT (", toSQL(p.right) + ")")
 
+    case p @ SubqueryAlias(_, Limit(_, _), _) => build(toSQL(p.child), "AS", p.alias)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: `SubqueryAlias` wraps the child by default. To prevent double `(( ))`, do not add '()' here.

**Comment by rxin** [comment]
cc @gatorsmile

**Comment by gatorsmile** [comment]
Will do the review soon. Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: This is a bug fix. It is not related to this PR. Could you create a separate one? Make it consistent with what we did in `SQLQueryTestSuite`? Thanks!

**Comment by gatorsmile** [comment]
@dongjoon-hyun I have a general comment about the SQL generation fixes. Not sure how you fix the issue. Normally, when we did it, we need to see the analyzed SQL plan. In this case, it becomes a little complex, we called `addSubqueryIfNeeded` in `toSQL`. 

Could you try to show reviewers the failed plans with the added `SubqueryAlias`? Then, we can discuss whether the fix coverage and test case coverage.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Sure. I'll make a PR for this.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile . I'll add the failed plans, too.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile . The PR is updated according to the advice.
- Remove the `goldenSQLPath` from this PR.
- Update the PR description about the target fixed cases and the detailed plans, which includes the `SubqueryAlias` cases.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile .
Could you review again?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala (hunk 7 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: double `((` `))` is no harm, we should prefer the simplicity of the generator code over the generated SQL.

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: Thank you for review, @cloud-fan . Yep. I'll remove this.
This will change a few generated SQLs by `LogicalPlanToSQLSuite` together.

**Comment by dongjoon-hyun** [comment]
During updating the PR, I rebased and squashed to resolve conflict.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: is this `case` really needed or just used to avoid duplicated `()`?

Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: This makes `((~) AS gen_subquery_0 LIMIT 2)` pattern.
For example, `select * from ((select 1) as T)` works, but `select * from ((select 1) as T limit 2)` fails.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: This is subtle, can you add some comments?

**Comment by dongjoon-hyun** [comment]
The three failures seems to be irrelevant.
```
[info] - randomized aggregation test - [typed, with partial + safe] - with grouping keys - with non-empty input *** FAILED *** (1 second, 419 milliseconds)
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: Sure.

**Comment by gatorsmile** [comment]
Will review this PR tonight. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile !

**Comment by dongjoon-hyun** [comment]
Could you review again, @gatorsmile ?

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.1.

**Comment by gatorsmile** [comment]
Could you please backport it to 2.0? It sounds the JIRA opener needs it in Spark 2.0 branch.

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .
I'll make a backport for 2.0, too.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @rxin and @cloud-fan , too.
